src/Bridge/Symfony/Bundle/Test/ApiTestCase.php

```diff
@@ -41,19 +41,24 @@ protected function tearDown(): void
 
     /**
      * Creates a Client.
+     * In case there is an already booted kernel, that kernel is used and therefore 
+     * the $kernelOptions are ignored.
      *
-     * @param array $kernelOptions  Options to pass to the createKernel method
+     * @param array $kernelOptions  Options to pass to the createKernel method if
+     *                              no kernel is booted before this call
      * @param array $defaultOptions Default options for the requests
      */
     protected static function createClient(array $kernelOptions = [], array $defaultOptions = []): Client
     {
-        $kernel = static::bootKernel($kernelOptions);
+        if (static::$kernel === null) {
```

Comment:
While reviewing I compared with the upstream `createClient` function from Symfony: https://github.com/symfony/symfony/blob/master/src/Symfony/Bundle/FrameworkBundle/Test/WebTestCase.php#L40

I may be missing something but it looks like Symfony never reuse an existing kernel.

Reply:
To be consistent with upstream is a strong argument, indeed. In any case that should be the final state.

Throwing an exception like upstream would have shown me my problem (although I would have to have read the code to understand, that createClient() also boots a kernel) - so it is a solution.

But in the other hand:
I am thinking of the following scenario:
A big test-class needs some test-entities. Maybe some variantions according to the individual tests. But that needs a kernel in any case. 
But also lets assume only a minority of the tests in that class need a client the rest only a kernel.

It depends how expensive the creation of the client is. If it is cheap - OK, we could always create a client to get a booted kernel (but that behaviour would then have to be documented better).
But if it is expensive to create a kernel it would be better to only boot the kernel when we do not need the client for the test. In that case we could not put the initialization of the entities in a `setUp()` method. We would have to call the correct variant of the initialization, one only with a kernel, one with a client, in every test.

My scenario might be not realistic as the way we usually structure our tests might imply that all tests of a class might need a client. But is it always the case?

But maybe the solution could be to also change upstream according to this solution here.

As far as I can see it would break no existing code. But it would open the option to create a kernel before the `createClient()` to do whatever has to be done before that call. (Which also would lead to a consitent behaviour between Symfony and Api-Platform ;-) )

Reply:
[Followup] If we decide to go the route of my proposal, I would like to add an exception in case the kerneloptions-paramter is not empty and a kernel is reused.

Reply:
I would prefer to keep consistent with the behavior of Symfony's `WebTestCase`. What I suggest is to throw the exception for now, and to open an issue on the Symfony bug tracker to propose your behavior. If it changes in SF, then we'll change it accordingly here.

`ApiTestCase` is designed to be as close as possible of `WebTestCase` and `PantherTestCase`.

WDYT?

Reply:
Agreed. Adding that exception like in `WebTestCase` is an improvement worth a change until upstream might be changed in favour of my solution which then has to implemented also here.
Will change that accordingly.

(So my first PR jumps to another project 8-) )

Reply:
@dunglas Unfortunately inserting the exception according to upstream breaks 7 existing tests:
```
ApiPlatform\Core\Tests\Bridge\Symfony\Bundle\Test\ApiTestCaseTest::testFindIriBy
ApiPlatform\Core\Tests\Bridge\Symfony\Bundle\Test\ClientTest::testRequest
ApiPlatform\Core\Tests\Bridge\Symfony\Bundle\Test\ClientTest::testCustomHeader
ApiPlatform\Core\Tests\Bridge\Symfony\Bundle\Test\ClientTest::testDefaultHeaders
ApiPlatform\Core\Tests\Bridge\Symfony\Bundle\Test\ClientTest::testAuthBasic with data set #0 ('dunglas:kevin')
ApiPlatform\Core\Tests\Bridge\Symfony\Bundle\Test\ClientTest::testAuthBasic with data set #1 (array('dunglas', 'kevin'))
ApiPlatform\Core\Tests\Bridge\Symfony\Bundle\Test\ClientTest::testComplexScenario
```
All these tests now throw the new LogicException because a kernel is booted before calling createClient.

IMHO changing those tests does not make much sense or might be difficult.

So suspending this PR and wait for the discussed change from upstream might be the way to go?